### PR TITLE
Improve `readline` typings

### PR DESCRIPTION
**Pull request by sobolevn**
Changes:
1. For some reason, this module was using old style `Optional`, is this a bug?
2. `set_startup_hook` can take any callable, the return type is ignored

### CODE REVIEW

**Comment by Akuli** [comment]
> this module was using old style Optional, is this a bug?

Type aliases still use old-style `Optional` and `Union`, because mypy had bugs where they didn't work in type aliases. There's still a bug that makes `_Color = str | tuple[float, float, float]` not work, and once that is fixed, we can apply the new syntax to type aliases too.

File: stdlib/readline.pyi [hunk 1]
```diff
@@ -1,10 +1,11 @@
 import sys
 from _typeshed import StrOrBytesPath
-from typing import Callable, Optional, Sequence
+from typing import Callable, Sequence
 
 if sys.platform != "win32":
-    _CompleterT = Optional[Callable[[str, int], Optional[str]]]
-    _CompDispT = Optional[Callable[[str, Sequence[str], int], None]]
+    _CompleterT = Callable[[str, int], str | None]
+    _CompDispT = Callable[[str, Sequence[str], int], None]
```
Comment (Akuli) [inline stdlib/readline.pyi:7]: Nit: These shouldn't be named ending with `T`, because they aren't type variables.

**Comment by AlexWaygood** [comment]
> 1. For some reason, this module was using old style `Optional`, is this a bug?

https://github.com/python/mypy/issues/11098 is the relevant mypy issue if you wanna help us get to a place where we can use PEP 604 syntax everywhere in typeshed ;) I tried looking at it today for a little while, but understanding the mypy codebase is still a little like wading through treacle for me :(

**Review by Akuli — APPROVED** [review]
Thanks!

**Comment by sobolevn** [comment]
Thanks for the review! I am trying to fix mypy to allow `tuple` aliases. See https://github.com/python/mypy/pull/12134

I hope it can make its way into `0.940`

**Pull request merged**